src/azul/service/manifest_service.py

```diff
@@ -714,11 +715,75 @@ def write_to(self, output: IO[str]) -> Optional[str]:
             # bundle UUID. Because a file can belong to multiple bundles we use
             # the one with the most recent version.
             bundle = max(doc['bundles'], key=itemgetter('version', 'uuid'))
-            output_name = bundle['uuid'] + '/' + name
+            output_name = self._sanitize_path(bundle['uuid'] + '/' + name)
             output.write(f'url={self._option(url.url)}\n'
                          f'output={self._option(output_name)}\n\n')
         return None
 
+    disallowed_characters = re.compile(r'[\x00-\x1f\\]')
+
+    replaceable_characters = re.compile(r'[<>:"|?*]')
+
+    valid_regex_sequence = re.compile(r'''
+        (\.?[^./ ])([^/]*[^./ ])? # sequences with '/', ' ', '.' zero or more times can compromise system files & paths,
+        (/(\.?[^./ ])([^/]*[^./ ])?)* # they are only allowed in between characters, not at start or end of a sequence
+    ''', re.X)
+
+    disallowed_path_components = {
+        'CON', 'PRN', 'AUX', 'NUL',
+        *(f'{cmd}{i}' for cmd in ['COM', 'LPT'] for i in range(1, 10))
+    }
+
+    @classmethod
+    def _sanitize_path(cls, path: str) -> str:
+        """
+        >>> f = CurlManifestGenerator._sanitize_path
+        >>> f('foo/bar/\\x1F/file') # doctest: +NORMALIZE_WHITESPACE
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid file name.', 'foo/bar/\\x1f/file',  'Control characters and backslash are not
+        allowed.', '\\x1f')
+
+        >>> f('foo/bar/\\/file') # doctest: +NORMALIZE_WHITESPACE
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid file name.', 'foo/bar/\\\\/file', 'Control characters and backslash are not
+        allowed.', '\\\\')
+
+        >>> f('foo/bar/COM6/file')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Path component not allowed', {'COM6'})
+
+        >>> f('foo/bar/ / baz/file')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Name sequence not allowed in file name', 'foo/bar/ / baz/file')
+
+        >>> f('foo/bar?/fi*le|')
+        'foo/bar_/fi_le_'
+
+        >>> f('foo/bar/file.fastqgz')
+        'foo/bar/file.fastqgz'
+        """
+        match = cls.disallowed_characters.search(path)
+        if match is not None:
+            raise RequirementError('Invalid file name.',
+                                   path,
+                                   'Control characters and backslash are not allowed.',
+                                   path[match.start()])
```

Comment:
Looks like you missed my reply to yours about this.

Reply:
I did, apologies. 
I couldn't come up with a reason for logging the index without providing more context.
I think however that there might be issues with a particular system printing out a Unicode character as oppose to an integer. But I don't know.

Reply:
If you log the character itself there is a great chance that it will be invisible or mangled in the logs, especially if it is a non-printable character or one with an ambiguous glyph (google that). Consider how many different kinds of dashes there are in unicode. Or types of horizontal spaces.